source/extensions/common/wasm/plugin.cc

```diff
@@ -20,6 +16,47 @@ WasmConfig::WasmConfig(const envoy::extensions::wasm::v3::PluginConfig& config)
     // TODO(rapilado): Set the SanitizationConfig fields once sanitization is implemented.
     allowed_capabilities_[capability.first] = proxy_wasm::SanitizationConfig();
   }
+
+  if (config_.vm_config().has_environment_variables()) {
+    auto envs = config_.vm_config().environment_variables();
+
+    // We reject NullVm with key_values configuration
+    // since it directly accesses Envoy's env vars and we should not modify Envoy's env vars here.
+    // TODO(mathetake): Once proxy_get_map_values(type::EnvironmentVariables, ..) call is supported,
+    // then remove this restriction
+    if (config.vm_config().runtime() == WasmRuntimeNames::get().Null &&
+        !envs.key_values().empty()) {
+      throw EnvoyException("envoy.extensions.wasm.v3.VmConfig.EnvironmentVariables.key_values must "
+                           "not be set for NullVm.");
+    }
+
+    // Check key duplication.
+    absl::flat_hash_set<std::string> keys;
+    for (auto& env : envs.key_values()) {
+      keys.insert(env.first);
+    }
+    for (auto& key : envs.host_env_keys()) {
+      if (keys.find(key) == keys.end()) {
+        keys.insert(key);
```

Comment:
Nit: you don't need to do the `find` and `insert`, just `insert` and look at the return value to see if the item was already there and the insert failed, see https://en.cppreference.com/w/cpp/container/map/insert